README


This archive contains the distribution MooseX-Types-Structured,
version 0.30:

  Structured Type Constraints for Moose

This software is copyright (c) 2008 by John Napiorkowski.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for MooseX-Types-Structured

0.30      2013-12-19 22:51:17Z
        - remove pod from internal and unindexed modules, for a prettier
          metacpan release page

0.29      2013-11-20 23:36:48Z
        - fixed use of an interface that was deprecated in Moose-2.1100
          (ether)
        - repository migrated from shadowcat to the github moose organization

0.28   2011-10-03
        - Fixed regression where mixed type constraints (MX:Types style and
        'classic' Stringy style) are used in a single structured type doesn't
        work.

0.27   2011-04-28
        - test fixes for new Moose

0.26   2011-01-02
        - removed version from Test::Fatal as asked by the debian folks
        - small documentation updates

0.25   2010-12-28
        - fixed bug where ->is_subtype_of dies meaninglessly when the type we
          are trying to check is not a type we can find.  This makes our
          handling consistent with core Moose.  Also changed ->equals and
          ->is_a_type_of to be consistent.
        - Added test case for above
        - The test suite now uses Test::Fatal instead of Test::Exception (Karen
          Etheridge).

0.24   2010-11-16
        - Added some performance enhancing caching code (phaeton)

0.23   2010-07-01
        - Changes to the return value of ->validate that hopefully is both
          backwardly compatible as well as more detailed.  Now if you have
          a deeply recursive of repeated type constraints inside other
          type constraints you will get a drill down report to show the
          actual type constraint that failed

0.22   2010-06-01
        - Added tests to demonstrate type constraint equality problem
          introduced in Moose 1.05

0.21   2010-03-26
        - Removed unneed module from test
        - additional contributed documentation fixes

0.20   2010-02-04
        - Add a new Map type. (Ricardo SIGNES)
        - Properly handle Optional[] types within Tuples and Dicts.
          (Florian Ragwitz)

0.19   2009-11-06
        - Require Devel::PartialDump 0.10 and change tests to expect the
          correct format of error.

0.18   2009-08-17
        - Changed the Makefile.PL to reduce dependencies not needed by users
          that are not authors.  Bumped the revision and released to clear an
          error with cpan permissions.

0.17   2009-08-12
        - No new functionality.
        - Changed the way we specify dependencies in the Makefile.PL to reduce
          the depth of the dependency chain in cases where we don't need to be
          running the author quality tests.
        - Some documentation tweaks.

0.16   2009-05-28
        - Fix failing tests and test warnings on MooseX::Types 0.11.

0.15   2009-05-27
        - Change copy on license and added contributors section

0.14   2009-05-01
        - Use a builder instead of wrapping new to set the default structured
          coercion (rafl).
        - Make overflow (slurpy) type constraints introspectable and the name
          of constraints using them reasonable (rafl).

0.13   2009-04-25
        - Explicitly don't inline yet another constructor to avoid
          warnings (autarch).

0.12   2009-04-21
        - Explicitly don't inline the constructor to avoid warnings (rafl).
        - Pathological test cases for API methods equals, is_a_type_of and
          is_subtype_of (rafl).
        - significant improvements to API methods is_a_type_of, is_subtype_of
          and equals (nothingmuch).

0.11   2009-04-06
        - Fixed braindead bug in the way I override ->validate, which caused
          valiate to never correctly pass a constraint.

0.10   2009-04-02
        - Minor documentation grammar fixes and one major example error fixed
        - Much improved error reporting.  Now we return the 'internal' error
          that kicked a validation failure.  It's still best to use this for
          debugging rather than for actual user error messages, since I think
          we are rapidly approaching a need for Moose constraints needing more
          in the error and message reporting.
        - Documentation for the above.
0.09   2009-03-07
        - I guess we don't support the "subtype MyType, [TypeConstraint]" syntax
          anymore.  Changed the recursion test to reflect that, which should fix
          my 100% fail rate :)
0.08   2009-03-06
        - New Feature: slurpy method for when you want a structured type
          constraint that allows trailing arguments of indeterminate length.
          Please see the documentation and the '11-overflow.t' test for more.
        - Documentation for above as well as a bunch of POD cleanups, spell
          checks and improvements to formatting.
        - Stevan Little submitted a sweet  update to the '10-recursion.t' test
          that blows my mind.  Will probably form the core of a to be done set
          of cookbook style PODs.  Worth looking at.
        - First step at improving the error message you get when validation
          fails.  A full error stacktrace is not in this release, but you now
          at least get to see part of the offending value.

0.07   2008-12-09
        - Fixed typo in previous changelog
        - documentation improvements and updates
        - increased version requirement for MooseX::Types so that we can take
          advantage of the recursion support added.
        - added test for recursion.

0.06   2008-12-06
        - Added a 'helper' type constraint called Optional. See docs for more.
        - added lots of tests to cover the API better, coverage and fixes for
          the ->parameterize method in particular have been clarified.
        - changes so that the type contraints are more forgiving when null
          values appear in method calls.
        - used ->make_immutable which should speed up the constraints.
        - removed some unnecessary calls to use Moose when I wasn't using Moose.

lib/MooseX/Types/Structured/MessageStack.pm
package ## Hide from PAUSE
  MooseX::Types::Structured::MessageStack;

use Moose;


has 'level' => (
    traits => ['Counter'],
    is => 'ro',
    isa => 'Num',
    required => 0,
    default => 0,
    handles => {
        inc_level => 'inc',
        dec_level => 'dec',
    },
);


has 'messages' => (
    traits => ['Array'],
    is => 'ro',
    isa => 'ArrayRef[HashRef]',
    required => 1,
    default => sub { [] },
    handles => {
        has_messages => 'count',
        add_message => 'push',
        all_messages => 'elements',
    },
);


sub as_string {
    my @messages = (shift)->all_messages;
    my @flattened_msgs =  map {
        "\n". (" " x $_->{level}) ."[+] " . $_->{message};
    } reverse @messages;

    return join("", @flattened_msgs);
}

no Moose;
__PACKAGE__->meta->make_immutable;

1;

__END__

=pod

=encoding UTF-8

=for :stopwords John Napiorkowski Florian Ragwitz יובל קוג'מן (Yuval Kogman) Tomas (t0m)
Doran Robert Sedlacek

=head1 NAME

MooseX::Types::Structured::MessageStack

=head1 VERSION

version 0.29

=head1 ATTRIBUTES

=head2 level

=head2 messages

=head1 METHODS

=head2 as_string

=head1 AUTHORS

=over 4

=item *

John Napiorkowski <jjnapiork@cpan.org>

=item *

Florian Ragwitz <rafl@debian.org>

=item *

יובל קוג'מן (Yuval Kogman) <nothingmuch@woobling.org>

=item *

Tomas (t0m) Doran <bobtfish@bobtfish.net>

=item *

Robert Sedlacek <rs@474.at>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2008 by John Napiorkowski.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/MooseX/Meta/TypeConstraint/Structured.pm
package ## Hide from PAUSE
 MooseX::Meta::TypeConstraint::Structured;
# ABSTRACT: Structured type constraints.

use Moose;
use Devel::PartialDump;
use Moose::Util::TypeConstraints ();
use MooseX::Meta::TypeCoercion::Structured;
extends 'Moose::Meta::TypeConstraint';



has 'type_constraints' => (
    is=>'ro',
    isa=>'Ref',
    predicate=>'has_type_constraints',
);


has 'constraint_generator' => (
    is=>'ro',
    isa=>'CodeRef',
    predicate=>'has_constraint_generator',
);

has coercion => (
    is      => 'ro',
    isa     => 'Object',
    builder => '_build_coercion',
);

sub _build_coercion {
    my ($self) = @_;
    return MooseX::Meta::TypeCoercion::Structured->new(
        type_constraint => $self,
    );
}


sub _clean_message {
    my $message = shift @_;
    $message =~s/MooseX::Types::Structured:://g;
    return $message;
}

override 'validate' => sub {
    my ($self, $value, $message_stack) = @_;
    unless ($message_stack) {
        $message_stack = MooseX::Types::Structured::MessageStack->new();
    }

    $message_stack->inc_level;

    if ($self->_compiled_type_constraint->($value, $message_stack)) {
        ## Everything is good, no error message to return
        return undef;
    } else {
        ## Whoops, need to figure out the right error message
        my $args = Devel::PartialDump::dump($value);
        $message_stack->dec_level;
        if($message_stack->has_messages) {
            if($message_stack->level) {
                ## we are inside a deeply structured constraint
                return $self->get_message($args);
            } else {
                my $message_str = $message_stack->as_string;
                return _clean_message($self->get_message("$args, Internal Validation Error is: $message_str"));
            }
        } else {
            return $self->get_message($args);
        }
    }
};


sub generate_constraint_for {
    my ($self, $type_constraints) = @_;
    return $self->constraint_generator->($self, $type_constraints);
}


sub parameterize {
    my ($self, @type_constraints) = @_;
    my $class = ref $self;
    my $name = $self->name .'['. join(',', map {"$_"} @type_constraints) .']';
    my $constraint_generator = $self->__infer_constraint_generator;

    return $class->new(
        name => $name,
        parent => $self,
        type_constraints => \@type_constraints,
        constraint_generator => $constraint_generator,
    );
}


sub __infer_constraint_generator {
    my ($self) = @_;
    if($self->has_constraint_generator) {
        return $self->constraint_generator;
    } else {
        return sub {
            ## I'm not sure about this stuff but everything seems to work
            my $tc = shift @_;
            my $merged_tc = [@$tc, @{$self->parent->type_constraints}];
            $self->constraint->($merged_tc, @_);
        };
    }
}


around 'compile_type_constraint' => sub {
    my ($compile_type_constraint, $self, @args) = @_;

    if($self->has_type_constraints) {
        my $type_constraints = $self->type_constraints;
        my $constraint = $self->generate_constraint_for($type_constraints);
        $self->_set_constraint($constraint);
    }

    return $self->$compile_type_constraint(@args);
};


around 'create_child_type' => sub {
    my ($create_child_type, $self, %opts) = @_;
    return $self->$create_child_type(
        %opts,
        constraint_generator => $self->__infer_constraint_generator,
    );
};


sub equals {
    my ( $self, $type_or_name ) = @_;
    my $other = Moose::Util::TypeConstraints::find_type_constraint($type_or_name)
      or return;

    return unless $other->isa(__PACKAGE__);

    return (
        $self->parent->equals($other->parent)
            and
        $self->type_constraints_equals($other)
    );
}

sub is_a_type_of {
    my ( $self, $type_or_name ) = @_;
    my $other = Moose::Util::TypeConstraints::find_type_constraint($type_or_name)
      or return;

    if ( $other->isa(__PACKAGE__) and @{ $other->type_constraints || [] }) {
        if ( $self->parent->is_a_type_of($other->parent) ) {
            return $self->_type_constraints_op_all($other, "is_a_type_of");
        } elsif ( $self->parent->is_a_type_of($other) ) {
            return 1;
            # FIXME compare?
        } else {
            return 0;
        }
    } else {
        return $self->SUPER::is_a_type_of($other);
    }
}

sub is_subtype_of {
    my ( $self, $type_or_name ) = @_;
    my $other = Moose::Util::TypeConstraints::find_type_constraint($type_or_name)
      or return;
    if ( $other->isa(__PACKAGE__) ) {
        if ( $other->type_constraints and $self->type_constraints ) {
            if ( $self->parent->is_a_type_of($other->parent) ) {
                return (
                    $self->_type_constraints_op_all($other, "is_a_type_of")
                      and
                    $self->_type_constraints_op_any($other, "is_subtype_of")
                );
            } elsif ( $self->parent->is_a_type_of($other) ) {
                return 1;
                # FIXME compare?
            } else {
                return 0;
            }
        } else {
            if ( $self->type_constraints ) {
                if ( $self->SUPER::is_subtype_of($other) ) {
                    return 1;
                } else {
                    return;
                }
            } else {
                return $self->parent->is_subtype_of($other->parent);
            }
        }
    } else {
        return $self->SUPER::is_subtype_of($other);
    }
}


sub type_constraints_equals {
    my ( $self, $other ) = @_;
    $self->_type_constraints_op_all($other, "equals");
}

sub _type_constraints_op_all {
    my ($self, $other, $op) = @_;

    return unless $other->isa(__PACKAGE__);

    my @self_type_constraints = @{$self->type_constraints||[]};
    my @other_type_constraints = @{$other->type_constraints||[]};

    return unless @self_type_constraints == @other_type_constraints;

    ## Incoming ay be either arrayref or hashref, need top compare both
    while(@self_type_constraints) {
        my $self_type_constraint = shift @self_type_constraints;
        my $other_type_constraint = shift @other_type_constraints;

        $_ = Moose::Util::TypeConstraints::find_or_create_isa_type_constraint($_)
          for $self_type_constraint, $other_type_constraint;

        my $result = $self_type_constraint->$op($other_type_constraint);
        return unless $result;
    }

    return 1; ##If we get this far, everything is good.
}

sub _type_constraints_op_any {
    my ($self, $other, $op) = @_;

    return unless $other->isa(__PACKAGE__);

    my @self_type_constraints = @{$self->type_constraints||[]};
    my @other_type_constraints = @{$other->type_constraints||[]};

    return unless @self_type_constraints == @other_type_constraints;

    ## Incoming ay be either arrayref or hashref, need top compare both
    while(@self_type_constraints) {
        my $self_type_constraint = shift @self_type_constraints;
        my $other_type_constraint = shift @other_type_constraints;

        $_ = Moose::Util::TypeConstraints::find_or_create_isa_type_constraint($_)
          for $self_type_constraint, $other_type_constraint;

        return 1 if $self_type_constraint->$op($other_type_constraint);
    }

    return 0;
}


around 'get_message' => sub {
    my ($get_message, $self, $value) = @_;
    $value = Devel::PartialDump::dump($value)
     if ref $value;
    return $self->$get_message($value);
};


__PACKAGE__->meta->make_immutable(inline_constructor => 0);

__END__

=pod

=encoding UTF-8

=for :stopwords John Napiorkowski Florian Ragwitz יובל קוג'מן (Yuval Kogman) Tomas (t0m)
Doran Robert Sedlacek subref parameterize servicable

=head1 NAME

MooseX::Meta::TypeConstraint::Structured - Structured type constraints.

=head1 VERSION

version 0.29

=head1 DESCRIPTION

A structure is a set of L<Moose::Meta::TypeConstraint> that are 'aggregated' in
such a way as that they are all applied to an incoming list of arguments.  The
idea here is that a Type Constraint could be something like, "An C<Int> followed by
an C<Int> and then a C<Str>" and that this could be done so with a declaration like:

    Tuple[Int,Int,Str]; ## Example syntax

So a structure is a list of type constraints (the C<Int,Int,Str> in the above
example) which are intended to function together.

=head1 ATTRIBUTES

=head2 type_constraints

A list of L<Moose::Meta::TypeConstraint> objects.

=head2 constraint_generator

=head1 METHODS

=head2 validate

Messing with validate so that we can support nicer error messages.

=head2 generate_constraint_for ($type_constraints)

Given some type constraints, use them to generate validation rules for an ref
of values (to be passed at check time)

=head2 parameterize (@type_constraints)

Given a ref of type constraints, create a structured type.

=head2 __infer_constraint_generator

=head2 compile_type_constraint

hook into compile_type_constraint so we can set the correct validation rules.

=head2 create_child_type

modifier to make sure we get the constraint_generator

=head2 is_a_type_of

=head2 is_subtype_of

=head2 equals

Override the base class behavior.

=head2 type_constraints_equals

Checks to see if the internal type constraints are equal.

=head2 get_message

Give you a better peek into what's causing the error.  For now we stringify the
incoming deep value with L<Devel::PartialDump> and pass that on to either your
custom error message or the default one.  In the future we'll try to provide a
more complete stack trace of the actual offending elements

A subref or closure that contains the way we validate incoming values against
a set of type constraints.

This returns a CODEREF which generates a suitable constraint generator.  Not
user servicable, you'll never call this directly.

=head1 SEE ALSO

The following modules or resources may be of interest.

L<Moose>, L<Moose::Meta::TypeConstraint>

=head1 AUTHORS

=over 4

=item *

John Napiorkowski <jjnapiork@cpan.org>

=item *

Florian Ragwitz <rafl@debian.org>

=item *

יובל קוג'מן (Yuval Kogman) <nothingmuch@woobling.org>

=item *

Tomas (t0m) Doran <bobtfish@bobtfish.net>

=item *

Robert Sedlacek <rs@474.at>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2008 by John Napiorkowski.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/MooseX/Types/Structured/OverflowHandler.pm
package ## Hide from PAUSE
  MooseX::Types::Structured::OverflowHandler;

use Moose;

use overload '""' => 'name', fallback => 1;


has type_constraint => (
    is       => 'ro',
    isa      => 'Moose::Meta::TypeConstraint',
    required => 1,
    handles  => [qw/check/],
);


sub name {
    my ($self) = @_;
    return 'slurpy ' . $self->type_constraint->name;
}

no Moose;
__PACKAGE__->meta->make_immutable;

1;

__END__

=pod

=encoding UTF-8

=for :stopwords John Napiorkowski Florian Ragwitz יובל קוג'מן (Yuval Kogman) Tomas (t0m)
Doran Robert Sedlacek

=head1 NAME

MooseX::Types::Structured::OverflowHandler

=head1 VERSION

version 0.29

=head1 ATTRIBUTES

=head2 type_constraint

=head1 METHODS

=head2 name

=head1 AUTHORS

=over 4

=item *

John Napiorkowski <jjnapiork@cpan.org>

=item *

Florian Ragwitz <rafl@debian.org>

=item *

יובל קוג'מן (Yuval Kogman) <nothingmuch@woobling.org>

=item *

Tomas (t0m) Doran <bobtfish@bobtfish.net>

=item *

Robert Sedlacek <rs@474.at>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2008 by John Napiorkowski.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/MooseX/Meta/TypeConstraint/Structured/Optional.pm
package ## Hide from PAUSE
  MooseX::Meta::TypeConstraint::Structured::Optional;

use Moose;
use MooseX::Meta::TypeCoercion::Structured::Optional;

extends 'Moose::Meta::TypeConstraint::Parameterizable';

around parameterize => sub {
    my $orig = shift;
    my $self = shift;

    my $ret = $self->$orig(@_);

    $ret->coercion(MooseX::Meta::TypeCoercion::Structured::Optional->new(type_constraint => $ret));

    return $ret;
};

__PACKAGE__->meta->make_immutable(inline_constructor => 0);

1;

__END__

=pod

=encoding UTF-8

=for :stopwords John Napiorkowski Florian Ragwitz יובל קוג'מן (Yuval Kogman) Tomas (t0m)
Doran Robert Sedlacek Ansgar 'phaylon' Stevan Little Yuval Kogman arcanez
Burchardt Dave Rolsky Jesse Luehrs Karen Etheridge Ricardo Signes

=head1 NAME

MooseX::Meta::TypeConstraint::Structured::Optional

=head1 VERSION

version 0.29

=head1 AUTHORS

=over 4

=item *

John Napiorkowski <jjnapiork@cpan.org>

=item *

Florian Ragwitz <rafl@debian.org>

=item *

יובל קוג'מן (Yuval Kogman) <nothingmuch@woobling.org>

=item *

Tomas (t0m) Doran <bobtfish@bobtfish.net>

=item *

Robert Sedlacek <rs@474.at>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2008 by John Napiorkowski.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL
# This Makefile.PL for MooseX-Types-Structured was generated by
# Dist::Zilla::Plugin::MakeMaker::Fallback 0.005.
# Don't edit it but the dist.ini used to construct it.

use strict;
use warnings;

BEGIN {
my %configure_requires = (
    'Module::Build::Tiny' => '0.030',
    'ExtUtils::MakeMaker' => '6.30',
);

my @missing = grep {
    ! eval "require $_; $_->VERSION($configure_requires{$_}); 1"
} keys %configure_requires;

if (not @missing)
{
    print "Congratulations, your toolchain understands 'configure_requires'!\n\n";
}
else
{
    $ENV{PERL_MM_FALLBACK_SILENCE_WARNING} or warn <<'EOW';
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

If you're installing manually, please retrain your fingers to run Build.PL
when present instead.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are doing something overly clever, in which case you
should consider setting the 'prefer_installer' config option in CPAN.pm, or
'prefer_makefile' in CPANPLUS, to 'mb" and '0' respectively.

You can also silence this warning for future installations by setting the
PERL_MM_FALLBACK_SILENCE_WARNING environment variable.

EOW

    sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));
}
}

use 5.008;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Structured Type Constraints for Moose",
  "AUTHOR" => "John Napiorkowski <jjnapiork\@cpan.org>, Florian Ragwitz <rafl\@debian.org>, \x{5d9}\x{5d5}\x{5d1}\x{5dc} \x{5e7}\x{5d5}\x{5d2}'\x{5de}\x{5df} (Yuval Kogman) <nothingmuch\@woobling.org>, Tomas (t0m) Doran <bobtfish\@bobtfish.net>, Robert Sedlacek <rs\@474.at>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build::Tiny" => "0.030"
  },
  "DISTNAME" => "MooseX-Types-Structured",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "MooseX::Types::Structured",
  "PREREQ_PM" => {
    "Devel::PartialDump" => "0.13",
    "Moose" => "1.08",
    "Moose::Meta::TypeCoercion" => 0,
    "Moose::Meta::TypeConstraint" => 0,
    "Moose::Meta::TypeConstraint::Parameterizable" => 0,
    "Moose::Util::TypeConstraints" => "1.06",
    "MooseX::Types" => "0.22",
    "Scalar::Util" => 0,
    "Sub::Exporter" => "0.982",
    "overload" => 0
  },
  "TEST_REQUIRES" => {
    "Data::Dumper" => 0,
    "DateTime" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec::Functions" => 0,
    "List::Util" => 0,
    "MooseX::Types::DateTime" => 0,
    "MooseX::Types::Moose" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "0.94",
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.30",
  "test" => {
    "TESTS" => "t/*.t t/regressions/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



